### PR TITLE
time for HGCAL realistic sim clusters

### PR DESCRIPTION
**Pull request by amartelli**
(reopening PR for merge in correct place)

Introduce timing for realistic sim clusters, in HGCAL only.
Use algorithm optimized for TDR studies.
N.B. time is referred to the start of the BX, it is corrected for tof in the hypothesis of neutral and light particle: for longer track length and beta < 1 need offset correction at analysis level

The timing for the hits in HGCAL is enabled in the PR #21957

@bendavid @rovere @felicepantaleo

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

File: RecoParticleFlow/PFClusterProducer/plugins/SimMappers/ComputeClusterTime.h
Comment (felicepantaleo) [inline RecoParticleFlow/PFClusterProducer/plugins/SimMappers/ComputeClusterTime.h]: do you need this?

File: RecoParticleFlow/PFClusterProducer/plugins/SimMappers/ComputeClusterTime.h
Comment (felicepantaleo) [inline RecoParticleFlow/PFClusterProducer/plugins/SimMappers/ComputeClusterTime.h]: replace `.at(i)` with `[i]` everywhere

File: RecoParticleFlow/PFClusterProducer/plugins/SimMappers/ComputeClusterTime.h
Comment (felicepantaleo) [inline RecoParticleFlow/PFClusterProducer/plugins/SimMappers/ComputeClusterTime.h]: evaluate `totSize*fractionToKeep` once and reuse the result

File: RecoParticleFlow/PFClusterProducer/plugins/SimMappers/ComputeClusterTime.h
Comment (felicepantaleo) [inline RecoParticleFlow/PFClusterProducer/plugins/SimMappers/ComputeClusterTime.h]: fix indentation here...

**Review by felicepantaleo — CHANGES_REQUESTED** [review]

**Comment by bendavid** [comment]
Since at the stage when pfcandidates are formed later we will want to correct the time to the primary vertex, it might be better at this stage to correct the time to the reconstructed cluster position (such that we avoid needing to propagate backwards and forwards for any subsequent vertex correction)

**Comment by amartelli** [comment]
@felicepantaleo changes implemented. Thanks!

**Comment by amartelli** [comment]
@bendavid 
At this stage the time is already corrected to the (0,0,0) vertex, so a propagation backwards is needed anyway. From my understanding having the time information corrected at the (0,0,0) vertex is a general convention for all subdetectors, so I am wondering if applying the backward propagation later, at the stage when you correct to the right primary vertex, can help the consistency and reduce any misinterpretation.  
Anyway I am not strongly against this, so if you think it's better I can correct the time back to the cluster here. (let me know)
cheers

**Comment by lgray** [comment]
@bendavid @amartelli (0,0,0) is used in the rechit and elsewhere in the digitization / local reconstruction since it reduces greatly the dynamic range of values that need to be stored in the digis or hits. FWIW it's really only needed in the digis to fit in a reasonable number of bits, and after that we may set the convention for RecHits or anything afterwards to be whatever is most convenient.

**Comment by perrotta** [comment]
@cmsbuild please test

**Comment by kpedro88** [comment]
assign upgrade

File: RecoParticleFlow/PFClusterProducer/plugins/SimMappers/ComputeClusterTime.h
Comment (kpedro88) [inline RecoParticleFlow/PFClusterProducer/plugins/SimMappers/ComputeClusterTime.h]: add an include guard for this header

File: RecoParticleFlow/PFClusterProducer/plugins/SimMappers/ComputeClusterTime.h
Comment (kpedro88) [inline RecoParticleFlow/PFClusterProducer/plugins/SimMappers/ComputeClusterTime.h]: put the functions in this header inside a namespace

File: RecoParticleFlow/PFClusterProducer/plugins/SimMappers/ComputeClusterTime.h
Comment (kpedro88) [inline RecoParticleFlow/PFClusterProducer/plugins/SimMappers/ComputeClusterTime.h]: this function isn't used anywhere?

File: RecoParticleFlow/PFClusterProducer/plugins/SimMappers/ComputeClusterTime.h
Comment (kpedro88) [inline RecoParticleFlow/PFClusterProducer/plugins/SimMappers/ComputeClusterTime.h]: don't mix tabs and spaces for indentation

File: RecoParticleFlow/PFClusterProducer/plugins/SimMappers/ComputeClusterTime.h
Comment (kpedro88) [inline RecoParticleFlow/PFClusterProducer/plugins/SimMappers/ComputeClusterTime.h]: this function and the previous function both sort a vector - if they end up sorting the same vector (in future usage), would be nice to isolate that step to avoid repeating it

File: RecoParticleFlow/PFClusterProducer/plugins/SimMappers/ComputeClusterTime.h
Comment (kpedro88) [inline RecoParticleFlow/PFClusterProducer/plugins/SimMappers/ComputeClusterTime.h]: can this just be a for loop?

File: RecoParticleFlow/PFClusterProducer/plugins/SimMappers/RealisticSimClusterMapper.cc
Comment (kpedro88) [inline RecoParticleFlow/PFClusterProducer/plugins/SimMappers/RealisticSimClusterMapper.cc]: no need to clear a vector just after declaring it

File: RecoParticleFlow/PFClusterProducer/plugins/SimMappers/ComputeClusterTime.h
Comment (amartelli) [reply RecoParticleFlow/PFClusterProducer/plugins/SimMappers/ComputeClusterTime.h]: No I will leave it commented out, as reference for future developments, and add a comment line about this.

File: RecoParticleFlow/PFClusterProducer/plugins/SimMappers/ComputeClusterTime.h
Comment (kpedro88) [inline RecoParticleFlow/PFClusterProducer/plugins/SimMappers/ComputeClusterTime.h]: Should be: `RecoParticleFlow_PFClusterProducer_ComputeClusterTime_h`

File: RecoParticleFlow/PFClusterProducer/plugins/SimMappers/ComputeClusterTime.h
Comment (kpedro88) [inline RecoParticleFlow/PFClusterProducer/plugins/SimMappers/ComputeClusterTime.h]: tabs and spaces for indentation are still mixed throughout this file and the next one.

File: RecoParticleFlow/PFClusterProducer/plugins/SimMappers/ComputeClusterTime.h
Comment (perrotta) [inline RecoParticleFlow/PFClusterProducer/plugins/SimMappers/ComputeClusterTime.h]: Although it is unlikely that there are two classes with the same name in CMSSW,  include guards should better include the full path or alike, e.g. ```RecoParticleFlow_PFClusterProducer_ComputeClusterTime_h```

File: RecoParticleFlow/PFClusterProducer/plugins/SimMappers/RealisticSimClusterMapper.cc
Comment (perrotta) [inline RecoParticleFlow/PFClusterProducer/plugins/SimMappers/RealisticSimClusterMapper.cc]: Please write the full path (for all three include's here)

File: RecoParticleFlow/PFClusterProducer/plugins/SimMappers/ComputeClusterTime.h
Comment (perrotta) [inline RecoParticleFlow/PFClusterProducer/plugins/SimMappers/ComputeClusterTime.h]: (Even if this part of the code is commented out)
No need to initialize startBin and endBin here: it can be done later at lines 110 and 111

File: RecoParticleFlow/PFClusterProducer/plugins/SimMappers/ComputeClusterTime.h
Comment (perrotta) [inline RecoParticleFlow/PFClusterProducer/plugins/SimMappers/ComputeClusterTime.h]: You already defined ```totToKeep```, you can just reuse it here and avoid further conversions

File: RecoParticleFlow/PFClusterProducer/plugins/SimMappers/ComputeClusterTime.h
Comment (perrotta) [inline RecoParticleFlow/PFClusterProducer/plugins/SimMappers/ComputeClusterTime.h]: no need of  ```int(...)```

**Comment by perrotta** [comment]
please test

File: RecoParticleFlow/PFClusterProducer/plugins/SimMappers/ComputeClusterTime.h
Comment (slava77) [inline RecoParticleFlow/PFClusterProducer/plugins/SimMappers/ComputeClusterTime.h]: the CMS style is for lower case letters for the namespaces
this is a common style in the industry as well

**Comment by kpedro88** [comment]
please test

**Comment by Saptaparna** [comment]
Workflow should be tested for 140 and/or 200 PU scenarios.

File: RecoParticleFlow/PFClusterProducer/plugins/SimMappers/RealisticSimClusterMapper.h
Comment (perrotta) [inline RecoParticleFlow/PFClusterProducer/plugins/SimMappers/RealisticSimClusterMapper.h]: The number of parameters is growing, and their names also get a bit more complicate.
I think that you can avoid future mistakes and problems if you switch since now to the fillDescription() method for them

**Comment by amartelli** [comment]
@perrotta  I understand the point, I looked for some info

With the converter suggested in  https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideConfigurationValidationAndHelp#Converting_an_Existing_CFI_file   
the produced function is void PFClusterProducer::fillDescriptions(...)
 
So, are you asking to add this obtained function  in 
https://github.com/cms-sw/cmssw/blob/master/RecoParticleFlow/PFClusterProducer/plugins/PFClusterProducer.cc  
and remove RecoParticleFlow/PFClusterProducer/python/particleFlowClusterHGC_cfi.py?

Or for this specific case, I can leave it as it is, since the particleFlowClusterHGC_cfi.py configuration file refers to many other classes than the simple RealisticSimClusterMapper, and probably the usage of fillDescription might be suggested independently of this specific addition of parameters?

thanks

File: RecoParticleFlow/PFClusterProducer/plugins/SimMappers/ComputeClusterTime.h
Comment (perrotta) [inline RecoParticleFlow/PFClusterProducer/plugins/SimMappers/ComputeClusterTime.h]: Are timeHits sorted? IOW, is it correct not to consider ```std::abs(el - startRef)``` here and below at L40?

**Comment by perrotta** [comment]
@amartelli : for the fillDescription() I was talking about RealisticSimClusterMapper
As you are pointing out, this is just a subset of the parameters of the larger particleFlowClusterHGC_cfi.py. While errors and inconsistencies in configs can be avoided by moving to fillDescription(), I understand that this move can be a change which goes beyond what you intend for this PR.
Take it as a suggestion for cleaner and more efficient code, and decide whether you want to apply it already in this PR, or defer to a possible future one.

**Comment by perrotta** [comment]
please test
(Just to recover the jenkins tests results which got removed in the meanwhile)

**Comment by perrotta** [comment]
@amartelli : the summary of the (previous) tests showed that there were no changes due to this PR, at least not amongst the monitored quantities.
Could you please summarize what should we expect in the outputs of the upgrade workflows as a result of the integration of this PR?

File: RecoParticleFlow/PFClusterProducer/plugins/SimMappers/ComputeClusterTime.h
Comment (amartelli) [reply RecoParticleFlow/PFClusterProducer/plugins/SimMappers/ComputeClusterTime.h]: yes timeHits are sorted

**Comment by amartelli** [comment]
@perrotta. The only expected difference is in the time of PFCluster. This quantity was meaningless (before this PR) so I won't be surprised if it is not amongst the monitored ones, and if this is the case, it's a good thing that no difference is spotted.

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
@amartelli : I run a couple of upgrade workflows (20034 with D17 detector and 20434 with D19 detector) in the baseline CMSSW_10_2_X_2018-04-15-2300 and in the same IB plus the addition of this PR. In both such workflows I see no changes at all in the HGCal PFCluster timing (both with and without MultiCluster). The example of the two distributions in wf 20034 is the following, and it is completely identical before and after the merging of this PR:
hgcaltime
hgcaltimefrommulticl

Since you claim in https://github.com/cms-sw/cmssw/pull/22758#issuecomment-381425907 that the time of PFCluster is supposed to change, could you please verify, and possibly post here exactly which time distributions are actually affected by this PR?

**Comment by amartelli** [comment]
@bendavid  Can you please comment about this (I don't have time to dig properly into the PF code right now and anyway you know what is going on with time and PF) 

I expected that PFCluster had no time assigned, but it seems it is not the case...

I made sure to setTime for PFCluster in https://github.com/amartelli/cmssw/blob/d1528323026c02b3076ad09f3678b689b7965c36/RecoParticleFlow/PFClusterProducer/plugins/SimMappers/RealisticSimClusterMapper.cc#L179
- as you suggested - but I guess that this gets overwritten somewhere else and ends up in the output collection. 

thanks for your info

**Comment by rovere** [comment]
hi all,
@perrotta , thanks for checking. In fact the timing is overwritten by the 3DPCACalculator, which is actually called for both collections (see here: [link](https://github.com/cms-sw/cmssw/blob/master/RecoParticleFlow/PFClusterProducer/plugins/Cluster3DPCACalculator.cc#L128))

Having said that, my naive expectations would have been:

* no timing info for both collections w/o this PR
* timing info only for `particleFlowClusterHGCAL` with this PR, none for the `FromMultiCl`.

My quick solution would be to disable the timing in the 3DPCA, either directly or using a flag (and add it if it's not there).

Timing people could maybe comment further.

**Comment by bendavid** [comment]
So I don't think we really need timing for the multiclusters at the moment, just the particleFlowClusterHGCAL which is used to make the PFCandidates.

Disabling the overwriting of the time in the 3DPCACalculator seems reasonable.

**Comment by perrotta** [comment]
@amartelli @rovere @bendavid : is there any follow up from your observations?

**Comment by rovere** [comment]
@perrotta not yet: I'd say monday next week we will provide a fix-commit.

**Comment by perrotta** [comment]
-1 
According to the authors the updates and the discussion continue in #23047

**Comment by perrotta** [comment]
@amartelli : since all further updates were applied to #23047, which is now fully signed by the way, could you please close this pull request?

**Comment by perrotta** [comment]
Github realized that the commits in this PR were just a subset of the ones of  #23047.
Therefore, when  #23047 got merged, also this one results as merged. This is not a problem, because the commits in common are identically the same for the two PRs: I'm just noticing it here for the record.

**Comment by fabiocos** [comment]
@perrotta thank you for the comment, just for clarity

**Pull request merged**
